workflow/snakemake_rules/common.smk

```diff
@@ -145,15 +152,39 @@ def _get_max_date_for_frequencies(wildcards):
             date.today() - offset
         )
 
-def _get_first(config, *keys):
-    """
-    Get the value of the first key in *keys* that exists in *config* and has a
-    non-empty value.
-
-    Raises a :class:`KeyError` if none of the *keys* exist with a suitable
-    value.
-    """
-    for key in keys:
-        if config.get(key) not in {"", None}:
-            return config[key]
-    raise KeyError(str(keys))
+def _get_upload_inputs(wildcards):
+    # The main workflow supports multiple inputs/origins, but our desired file
+    # structure under data.nextstrain.org/files/ncov/open/… is designed around
+    # a single input/origin.  Intermediates (aligned, masked, filtered, etc)
+    # are specific to each input/origin and thus do not match our desired
+    # structure, while builds (global, europe, africa, etc) span all
+    # inputs/origins (and thus do).  In our desired outcome, the two kinds of
+    # files are comingled.  Without changing the main workflow, the mismatch
+    # has to be reconciled by the upload rule.  Thus, the upload rule enforces
+    # and uses only single-origin configurations.  How third-wave.
+    #   -trs, 7 May 2021
+    if len(config["S3_DST_ORIGINS"]) != 1:
+        raise Exception(f'The "upload" rule requires a single value in S3_DST_ORIGINS (got {config["S3_DST_ORIGINS"]!r}).')
+
+    origin = config["S3_DST_ORIGINS"][0]
+
+    # mapping of remote → local filenames
+    uploads = {
+        f"aligned.fasta.xz":              f"results/aligned_{origin}.fasta.xz",              # from `rule align`
+        f"masked.fasta.xz":               f"results/masked_{origin}.fasta.xz",               # from `rule mask`
+        f"filtered.fasta.xz":             f"results/filtered_{origin}.fasta.xz",             # from `rule filter`
+        f"mutation-summary.tsv.xz":       f"results/mutation_summary_{origin}.tsv.xz",       # from `rule mutation_summary`
+    }
+
+    for build_name in config["builds"]:
+        uploads.update({
+            f"{build_name}/sequences.fasta.xz": f"results/{build_name}/{build_name}_subsampled_sequences.fasta.xz",   # from `rule combine_samples`
+            f"{build_name}/metadata.tsv.xz":    f"results/{build_name}/{build_name}_subsampled_metadata.tsv.xz",      # from `rule combine_samples`
+            f"{build_name}/aligned.fasta.xz":   f"results/{build_name}/aligned.fasta.xz",                             # from `rule build_align`
+            # export the auspice dataset which matches the subsampled sequences / metadata (see `rule finalize`)
+            f"{build_name}/{build_name}.json":                  f"auspice/{config['auspice_json_prefix']}_{build_name}.json",
```

Comment:
The trial-build fails at this line:

```
[33m./scripts/upload-to-s3 auspice/ncov_open_trial_2021-06-26_global.json s3://nextstrain-staging/files/ncov/open/trial/2021-06-26/global/global.json \| tee -a logs/upload.txt[0m
--
An error occurred (404) when calling the HeadObject operation: Not Found
054e9d3a96a701a3db508e0af7030e9b2c257fa2eb4d5dbf72bbf043c2c7979b auspice/ncov_open_trial_2021-06-26_global.json
0000000000000000000000000000000000000000000000000000000000000000 s3://nextstrain-staging/files/ncov/open/trial/2021-06-26/global/global.json
Uploading auspice/ncov_open_trial_2021-06-26_global.json → s3://nextstrain-staging/files/ncov/open/trial/2021-06-26/global/global.json
Unknown options:
[32m[Fri Jun 25 06:24:43 2021][0m
[31mError in rule upload:[0m
[31m    jobid: 0[0m
[31m    log: logs/upload.txt (check log file(s) for error message)[0m
[31m[0m
[31mRuleException:
CalledProcessError in line 214 of /nextstrain/build/workflow/snakemake_rules/export_for_nextstrain.smk:
Command 'set -euo pipefail;  ./scripts/upload-to-s3 auspice/ncov_open_trial_2021-06-26_global.json s3://nextstrain-staging/files/ncov/open/trial/2021-06-26/global/global.json \| tee -a logs/upload.txt' returned non-zero exit status 255.
File "/nextstrain/build/workflow/snakemake_rules/export_for_nextstrain.smk", line 214, in __rule_upload
File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 57, in run[0m
[31mExiting because a job execution failed. Look above for error message[0m
```

_This particular line may not be the bug, but this is the part of the code where the error originates_

Reply:
Fixed by 2556b61